packages/opentelemetry-tracing/src/Tracer.ts

```diff
@@ -127,6 +127,88 @@ export class Tracer implements api.Tracer {
     return span;
   }
 
+  /**
+   * Starts a new {@link Span} and calls the given function passing it the
+   * created span as first argument.
+   * Additionally the new span gets set in context and this context is activated
+   * for the duration of the function call.
+   *
+   * @param name The name of the span
+   * @param [options] SpanOptions used for span creation
+   * @param [context] Context to use to extract parent
+   * @param fn function called in the context of the span and receives the newly created span as an argument
+   * @returns return value of fn
+   * @example
+   *   const something = tracer.startActiveSpan('op', span => {
+   *     try {
+   *       do some work
+   *       span.setStatus({code: SpanStatusCode.OK});
+   *       return something;
+   *     } catch (err) {
+   *       span.setStatus({
+   *         code: SpanStatusCode.ERROR,
+   *         message: err.message,
+   *       });
+   *       throw err;
+   *     } finally {
+   *       span.end();
+   *     }
+   *   });
+   * @example
+   *   const span = tracer.startActiveSpan('op', span => {
+   *     try {
+   *       do some work
+   *       return span;
+   *     } catch (err) {
+   *       span.setStatus({
+   *         code: SpanStatusCode.ERROR,
+   *         message: err.message,
+   *       });
+   *       throw err;
+   *     }
+   *   });
+   *   do some more work
+   *   span.end();
+   */
+   startActiveSpan<F extends (span: api.Span) => ReturnType<F>>(
+    name: string,
+    arg2: F | api.SpanOptions,
+    arg3?: F | api.Context,
+    arg4?: F
+  ): ReturnType<F> | undefined {
+    let fn: F | undefined,
+      options: api.SpanOptions | undefined,
+      activeContext: api.Context | undefined;
+    if (arguments.length === 2 && typeof arg2 === 'function') {
+      fn = arg2;
+    } else if (
+      arguments.length === 3 &&
+      typeof arg2 === 'object' &&
+      typeof arg3 === 'function'
+    ) {
+      options = arg2;
+      fn = arg3;
+    } else if (
+      arguments.length === 4 &&
+      typeof arg2 === 'object' &&
+      typeof arg3 === 'object' &&
+      typeof arg4 === 'function'
+    ) {
+      options = arg2;
+      activeContext = arg3;
+      fn = arg4;
+    }
```

Comment:
No need for typeof checking. There's not much we can do if the user provides us with wrong types and typeof isn't that accurate anyways (null is typeof object, etc).

Giving args good names and overwriting them makes it a little more clear IMO.

Options can be undefined if you want to provide a context but not any span options: `startActiveSpan("name", undefined, ctx, fn)`

```typescript
  startActiveSpan<F extends (span: api.Span) => ReturnType<F>>(
    name: string,
    options?: F | api.SpanOptions,
    parentContext?: F | api.Context,
    fn?: F
  ): ReturnType<F> | undefined {
  	if (arguments.length < 2) {
  	  return;
  	} else if (arguments.length === 2) {
      fn = options;
      options = undefined;
    } else if (arguments.length === 3) {
      fn = activeContext;
      parentContext = undefined;
    }
```

Reply:
> Giving args good names and overwriting them makes it a little more clear IMO.

what is meant by overwriting them?


apart from that yes agreed, PTAL though I am now faced with a new issue since applying code review

> This overload signature is not compatible with its implementation signature.ts(2394)

Reply:
your refactor doesn't correctly do overloads. you may want to look into the typescript documentation

Reply:
will do